File: solutions/python/anagram/1/anagram.py

def is_anagram(word, candidate):
    word, candidate = word.lower(), candidate.lower()
    if word == candidate:
        return False
    if len(word) != len(candidate):
        return False
    new_word, new_candidate = tuple(sorted(word)), tuple(sorted(candidate))
    return all( nw == nc for nw, nc in zip(new_word, new_candidate))


def find_anagrams(word, candidates):
    return [ item for item in candidates if is_anagram(word, item)]

File: solutions/java/anagram/1/src/main/java/Anagram.java
import java.lang.Character;
import java.lang.String;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class Anagram {

    private final List<Character> wordChars;
    private final String word;

    public Anagram(String word) {
        this.word = word.toLowerCase();
        this.wordChars = this.word.chars().sorted()
                             .mapToObj(c -> (char) c).collect(Collectors.toList());
    }

    public boolean isAnagram(String candidate) {
        if (candidate == null) { return false; }
        if (candidate.toLowerCase().equals(this.word)) { return false; }
        return candidate.toLowerCase().chars().sorted()
                        .mapToObj(c -> (char) c)
                        .collect(Collectors.toList()).equals(this.wordChars);
    }

    public List<String> match(List<String> candidates) {
        return IntStream.range(0, candidates.size())
                .filter(i -> isAnagram(candidates.get(i)))
                .mapToObj(candidates::get)
                .collect(Collectors.toList());
    }
}
